### ISSUE TITLE
Generate comparable FERC & EIA feature vectors

### ISSUE THREAD

**Issue by zaneselvans** [catalyst-cooperative/rmi-ferc1-eia#7]
Now that we're close to having reproduced the EIA Master Unit List, using data from across the entire range of available years, we can look at comparing those aggregated records with the same data as reported in FERC plant tables.

One way to go about doing this is to generate numerical feature vectors for each FERC and (aggregated) EIA record, and then within each year of reported data, for each FERC feature vector, find the most similar EIA feature vector.  This is a very general and flexible, but mathy approach.

**[Open Refine](https://openrefine.org/documentation.html)** is another toolkit that may be useful, but doesn't immediately jump out as the right choice for this particular task.

**Some background resources:**
* Note that pandas and scikit-learn have been much better integrated since the release of scikit-learn v0.20.0 (2018-09-25). Beware of older blog posts and documentation that describe what look like really painful ways to go about doing e.g. categorical encoding. -- it's much easier now. Unfortunately most of the examples floating around seem to date from that earlier period. Also note a relatively major release (v0.22.0) is about to come out, but looking at the **[change log](https://scikit-learn.org/0.22/whats_new/v0.22.html)** it doesn't appear there's anything major that will affect us.
* Concepts behind using **[Term frequency inverse document frequency](https://janav.wordpress.com/2013/10/27/tf-idf-and-cosine-similarity/)** (TF-IDF) and cosine similarity together to vectorize textual data and match similar strings. The discussion of cosine similarity is good, however, I'm not sure that the plant and utility names will be super useful in our case, since they have already gotten numerical IDs assigned to them. So we may be able to dispense with the TF-IDF stuff. This is good because textual data quickly blows up the size of the feature vectors (also here's **[another look at optimizing](https://bergvca.github.io/2017/10/14/super-fast-string-matching.html
)** this approach)
* **[This is a good video](https://www.youtube.com/watch?list=PL5-da3qGB5ICeMbQuqbbCOQWcS6OYBr5A&v=irHhDMbw3xo&feature=emb_title)** talking about how to use recent versions of pandas and scikit-learn together in order to transform a pandas dataframe into a feature matrix suitable for use in a wide variety of machine learning contexts -- especially the encoding of categorical values.  The older videos in the playlist are probably useful conceptually, but also use older versions of the software that didn't play as well together, and go well outside of the things I think we'll need to do to match up the FERC and EIA records.
* **[Cosine similarity](https://en.wikipedia.org/wiki/Cosine_similarity)** is an n-dimensional measure of feature vector similarity, which we can use to very rapidly compare FERC records to all (or a subset of) the EIA records and find the best match.
* Another **[recent and straightforward example](https://www.mikulskibartosz.name/preprocessing-the-input-pandas-dataframe-using-columntransformer-in-scikit-learn/)** of creating a scikit-learn ColumnTransformer with pandas, including a OneHotEncoder for categorical values, as well as some numerical data.
* We'll also want to assign weights to the features that are coming out of the ColumnTransformer, and ideally we'll be able to allow the computer to vary those weights to find the combination that gives the best matching outcomes using GridSearch and cross validation.

**Shared FERC / EIA Features**
* plant name (probably not required because we have the PUDL plant IDs)
* PUDL plant ID (categorical)
* utility / operator / respondent / owner name (also probably superceded by the PUDL utility ID)
* PUDL utility ID (categorical)
* plant/unit/generator capacity (MW) (numeric)
* net generation (MWh) (numeric)
* total cost of natural gas consumed ($) (numeric)
* total cost of petroleum consumed ($) (numeric)
* total cost of coal consumed ($) (numeric)
* total heat content of natural gas consumed (mmbtu)  (numeric)
* total heat content of petroleum consumed (mmbtu)  (numeric)
* total heat content of coal consumed (mmbtu) (numeric)
* construction year (first year a unit was installed) (categorical)
* installation (year of installation of most recently installed unit) (categorical)
* **anything else we should add???**

**Partitioning the matching search space**
* We have to partition by year (it would make no sense to match records from different years to each other).
* We can also partition by `plant_id_pudl` which should make the whole process extremely fast and memory efficient.  But we could also leave this out to begin with just to check for possible Plant ID errors.

**Other Questions**
* Do we want to allow the re-use of EIA records in the matching? I.e. is it acceptable for the same EIA record to end up matching with more than a single FERC record?
* Many of the features we have available don't exist for all the different types of plants (steam, small, hydro... what else?) How do we want to deal with those NA values?  Different ColumnTransformers and matching runs for each type of plant, with different feature vectors?
* We will need to combine the `plants_steam_ferc1` and `fuel_ferc1` tables to get more granular fuel data into the features for matching. Is the existing `fbp_ferc1` output sufficient?

**Metrics of success?**
* Can we pull the set of records that you've matched by hand for 2018 and use them as a training dataset (i.e. the matches that you've produced should be accurately reproduced by any automatic matching method we use). If so then we should be able to automatically allow the system to choose a good set of feature weights, based on its ability to reproduce these outputs (assuming they include a decent cross-section of the overall dataset).
* Can we construct some metric of time-series continuity that indicates whether the collections of FERC records and EIA plant parts are relatively consistent across years? Usually they should be -- if it's a big random mess we're doing something wrong.
* **What other post-matching sanity checking / metrics can we construct?**
